src/mgmt/srv_drpc.c

```diff
@@ -1803,6 +1803,90 @@ ds_mgmt_drpc_dev_set_faulty(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 	D_FREE(resp);
 }
 
+void
+ds_mgmt_drpc_dev_replace(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
+{
+	struct drpc_alloc	alloc = PROTO_ALLOCATOR_INIT(alloc);
+	Mgmt__DevReplaceReq	*req = NULL;
+	Mgmt__DevReplaceResp	*resp = NULL;
+	uint8_t			*body;
+	size_t			 len;
+	uuid_t			 old_uuid;
+	uuid_t			 new_uuid;
+	int			 rc = 0;
+
+	/* Unpack the inner request from the drpc call body */
+	req = mgmt__dev_replace_req__unpack(&alloc.alloc,
+					    drpc_req->body.len,
+					    drpc_req->body.data);
+
+	if (alloc.oom || req == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to unpack req (dev state set faulty)\n");
+		return;
+	}
+
+	D_INFO("Received request to replace device %s with device %s\n",
+	       req->old_dev_uuid, req->new_dev_uuid);
+
+	D_ALLOC_PTR(resp);
+	if (resp == NULL) {
+		drpc_resp->status = DRPC__STATUS__FAILURE;
+		D_ERROR("Failed to allocate daos response ref\n");
+		mgmt__dev_replace_req__free_unpacked(req, &alloc.alloc);
+		return;
+	}
+
+	/* Response status is populated with SUCCESS on init. */
+	mgmt__dev_replace_resp__init(resp);
+
+	if (strlen(req->old_dev_uuid) != 0) {
```

Comment:
Is this strlen check necessary? If the uuid is empty, doesn't `uuid_parse()` fail? It seems kind of strange to allow an empty uuid string and then catch it later as a null UUID in `ds_mgmt_dev_replace()`. I agree that `ds_mgmt_dev_replace()` should be doing its own sanity checking, but it seems like it would be clearer for this function to handle errors like that locally.

Reply:
ok I will add a check for this earlier and fail if either the new device or the old device uuid is empty